### PR TITLE
Fix haskell_toolchain invalidation bug

### LINKED ISSUE #980: haskell_register_ghc_nixpkgs incorrectly invalidated
**Describe the bug**

`haskell_register_ghc_nixpkgs` is not fully reproductible. If something changes in the setup of the rule, it may not force a rebuild of the associated rules.

**To Reproduce**

From master, I'm adding the `MemoTrie` package to the default toolchain using the following:

```
[nix-shell:~/tweag/rules_haskell]$ git diff
diff --git a/WORKSPACE b/WORKSPACE
index fcb1b79..20a1f2b 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -142,7 +142,8 @@ load(
 )
 
 haskell_register_ghc_nixpkgs(
-    attribute_path = "ghc",
+    attribute_path = "",
+    nix_file_content = "with import <nixpkgs> {}; haskellPackages.ghcWithPackages(p: [ p.MemoTrie])",
     compiler_flags = test_compiler_flags,
     haddock_flags = test_haddock_flags,
     locale_archive = "@glibc_locales//:locale-archive",
diff --git a/tests/hackage/BUILD.bazel b/tests/hackage/BUILD.bazel
index decba3f..87347a2 100644
--- a/tests/hackage/BUILD.bazel
+++ b/tests/hackage/BUILD.bazel
@@ -31,5 +31,6 @@ load(
         "process",
         "rts",
         "text",
+        "MemoTrie",
     ]
 ]
```

For the test, start by "not adding" `p.MemoTrie` in the `ghcWithPackages` and then build `//tests/hackage:MemoTrie`, it fails as expected:

```
[nix-shell:~/tweag/rules_haskell]$ bazel build //tests/hackage:MemoTrie
Starting local Bazel server and connecting to it...
/nix/store/gkpdg9y4420znmds66hlqzchv02swvgp-bazel-cc-toolchain
/nix/store/8b732fcj71j3mxp99g8104q9r88m5can-ghc-8.6.5
/nix/store/fnr9mys8l5224sqc7vgisgb66s5429f4-glibc-locales-2.27
ERROR: /home/guillaume/tweag/rules_haskell/tests/hackage/BUILD.bazel:17:5: in haskell_toolchain_library rule //tests/hackage:MemoTrie: 
Traceback (most recent call last):
	File "/home/guillaume/tweag/rules_haskell/tests/hackage/BUILD.bazel", line 17
		haskell_toolchain_library(name = 'MemoTrie')
	File "/home/guillaume/.cache/bazel/_bazel_guillaume/80819d925a91d7b49ed839165b536ee7/external/io_tweag_rules_haskell/haskell/private/haskell_impl.bzl", line 625, in haskell_toolchain_library_impl
		fail("\n{} is not a toolchain library...))

MemoTrie is not a toolchain library.
Check that it ships with your version of GHC.
            
ERROR: Analysis of target '//tests/hackage:MemoTrie' failed; build aborted: Analysis of target '//tests/hackage:MemoTrie' failed; build aborted
INFO: Elapsed time: 7.437s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (31 packages loaded, 614 targets configured)
```

Immediately after, add the `p.MemoTrie` in the `ghcWithPackage` list, it will fail similarly. `bazel clean` does not solve the issue, but `bazel clean --expunge` does:

```
[nix-shell:~/tweag/rules_haskell]$ cat WORKSPACE  | grep MemoTrie
    nix_file_content = "with import <nixpkgs> {}; haskellPackages.ghcWithPackages(p: [ p.MemoTrie])",
```

```
[nix-shell:~/tweag/rules_haskell]$ bazel build //tests/hackage:MemoTrie
/nix/store/rj0yma406y01q16djzabdwcs2cm2zv6h-ghc-8.6.5-with-packages
ERROR: /home/guillaume/tweag/rules_haskell/tests/hackage/BUILD.bazel:17:5: in haskell_toolchain_library rule //tests/hackage:MemoTrie: 
Traceback (most recent call last):
	File "/home/guillaume/tweag/rules_haskell/tests/hackage/BUILD.bazel", line 17
		haskell_toolchain_library(name = 'MemoTrie')
	File "/home/guillaume/.cache/bazel/_bazel_guillaume/80819d925a91d7b49ed839165b536ee7/external/io_tweag_rules_haskell/haskell/private/haskell_impl.bzl", line 625, in haskell_toolchain_library_impl
		fail("\n{} is not a toolchain library...))

MemoTrie is not a toolchain library.
Check that it ships with your version of GHC.
            
ERROR: Analysis of target '//tests/hackage:MemoTrie' failed; build aborted: Analysis of target '//tests/hackage:MemoTrie' failed; build aborted
INFO: Elapsed time: 2.566s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (2 packages loaded, 100 targets configured)
```
```
[nix-shell:~/tweag/rules_haskell]$ bazel clean --expunge
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
```

```
[nix-shell:~/tweag/rules_haskell]$ bazel build //tests/hackage:MemoTrie
Starting local Bazel server and connecting to it...
/nix/store/gkpdg9y4420znmds66hlqzchv02swvgp-bazel-cc-toolchain
/nix/store/rj0yma406y01q16djzabdwcs2cm2zv6h-ghc-8.6.5-with-packages
/nix/store/fnr9mys8l5224sqc7vgisgb66s5429f4-glibc-locales-2.27
INFO: Analyzed target //tests/hackage:MemoTrie (31 packages loaded, 629 targets configured).
INFO: Found 1 target...
Target //tests/hackage:MemoTrie up-to-date (nothing to build)
INFO: Elapsed time: 7.691s, Critical Path: 0.01s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
```

**Expected behavior**

This should not happen and changes to the `haskell_register_ghc_nixpkgs` arguments should be taken into account without `bazel clean --expunge`.

**Environment**
 - OS name + version: nixos (unstable)
 - Bazel version: 0.27 (rules_haskell master)
 - Version of the rules: master.

**Additional context**

This call (generating the `BUILD` file with the list of all dependencies) seems not to be invalidated:

https://github.com/tweag/rules_haskell/blob/master/haskell/nixpkgs.bzl#L43

### PR DESCRIPTION
**Pull request by mboes**
Fixes #980

Unfortunately, repositories cannot depend on each other directly. They
can only depend on files inside each repository. We need to be careful
to depend on files that change anytime any content in a repository
changes, otherwise repository rules won't be rerun every time an
upstream repository rule gets rerun.

### CODE REVIEW

**Comment by guibou** [comment]
I don't really understand how this works.

**Comment by guibou** [comment]
Actually, I just answered how do you ensure that `@io_tweag_rules_haskell_ghc_nixpkgs//:bin/ghc` is enough: By the fact that it contains path in nix store which are invalidated if the derivation changes.

Could you detail your comment about that?

File: haskell/nixpkgs.bzl [hunk 1]
```diff
@@ -96,7 +96,11 @@ _ghc_nixpkgs_haskell_toolchain = repository_rule(
         "haddock_flags": attr.string_list(),
         "repl_ghci_args": attr.string_list(),
         "locale_archive": attr.string(),
-        "_nixpkgs_ghc": attr.label(default = "@io_tweag_rules_haskell_ghc_nixpkgs//:BUILD"),
+        # Unfortunately, repositories cannot depend on each other
+        # directly. They can only depend on files inside each
+        # repository. We need to be careful to depend on files that
+        # change anytime any content in a repository changes.
+        "_nixpkgs_ghc": attr.label(default = "@io_tweag_rules_haskell_ghc_nixpkgs//:bin/ghc"),
```
Comment (guibou) [inline haskell/nixpkgs.bzl]: Please add a comment explaining why `bin/ghc` always change, contrary to `BUILD`. It took me a few time to understand why this is correct.

File: haskell/nixpkgs.bzl [hunk 2]
```diff
@@ -12,7 +12,7 @@ def _ghc_nixpkgs_haskell_toolchain_impl(repository_ctx):
         },
     )
     locale_archive = repository_ctx.attr.locale_archive
-    nixpkgs_ghc_path = repository_ctx.path(repository_ctx.attr._nixpkgs_ghc).dirname
+    nixpkgs_ghc_path = repository_ctx.path(repository_ctx.attr._nixpkgs_ghc).dirname.dirname
```
Comment (guibou) [inline haskell/nixpkgs.bzl]: Why?

File: haskell/nixpkgs.bzl (hunk 1 above)
Comment (mboes) [reply haskell/nixpkgs.bzl]: There's a comment just above. I'm not sure what more you want to see. A description of why `bin/ghc` changes?

File: haskell/nixpkgs.bzl (hunk 2 above)
Comment (mboes) [reply haskell/nixpkgs.bzl]: `/BUILD` vs `/bin/ghc`.

File: haskell/nixpkgs.bzl (hunk 1 above)
Comment (guibou) [reply haskell/nixpkgs.bzl]: Perhaps that's a nitpick, but yes.

**Pull request merged**